### PR TITLE
Make decoding jittable.

### PR DESCRIPTION
**Pull request by adarob**
Also returns actual logits instead of log probabilities.

### CODE REVIEW

**Comment by adarob** [comment]
Note that I have also implemented support for the following features, which I'd be happy to add:
  * Stacked LSTM
  * Bidirectional encoder
  * Variational loss (Recurrent VAE)
  * Scheduled sampling
  * Sampling w/ temperature

**Comment by avital** [comment]
@adarob that's super awesome! can we make those other modifications into [HOWTOs](https://flax.readthedocs.io/en/latest/howtos.html)? Happy to set up a call with you to work through the details though that doc should do a pretty decent job explaining the system.

File: examples/seq2seq/train.py [hunk 1]
```diff
@@ -156,17 +155,19 @@ def apply(self,
...
+    output = []
+    for _ in range(max_output_len):
+      next_inputs = next_inputs[:, np.newaxis]
+      carry, decoder_outputs = decoder(carry, next_inputs)
+      decoder_outputs = decoder_outputs.squeeze()
+      output.append(decoder_outputs)
+      # Select the argmax as the next input.
+      next_inputs = jnp.equal(
+          decoder_outputs,
+          jnp.max(decoder_outputs, axis=-1)[:, None]
```
Comment (marcvanzee) [inline examples/seq2seq/train.py]: Just for my understanding: axis=1 would also work here, right? What the intuition behind choosing -1?

File: examples/seq2seq/train.py [hunk 2]
```diff
@@ -156,17 +155,19 @@ def apply(self,
...
-    return output
+    next_inputs = jnp.zeros((batch_size, vocab_size))
+    output = []
+    for _ in range(max_output_len):
+      next_inputs = next_inputs[:, np.newaxis]
+      carry, decoder_outputs = decoder(carry, next_inputs)
+      decoder_outputs = decoder_outputs.squeeze()
+      output.append(decoder_outputs)
+      # Select the argmax as the next input.
+      next_inputs = jnp.equal(
```
Comment (marcvanzee) [inline examples/seq2seq/train.py]: Nice improvement! :)

File: examples/seq2seq/train.py (hunk 1 above)
Comment (adarob) [reply examples/seq2seq/train.py]: That's correct. I typically like to use axis=-1 in these cases since that is _almost always_ the logit/label dimension whereas sometimes dimension 1 is batch, sometimes it is sequence position, etc. Of course, we should try to be consistent with this.

File: examples/seq2seq/train.py [hunk 3]
```diff
@@ -251,14 +252,20 @@ def log_decode(question, inferred, golden):
   logging.info('DECODE: %s = %s %s', question, inferred, suffix)
 
 
+@jax.jit
+def decode(model, inputs):
+  """Decode inputs."""
+  decoder_inputs = encode_onehot(jnp.array(['='])).squeeze()
```
[the remark this answers is not in the document]
Comment (adarob) [reply examples/seq2seq/train.py]: Done.
